File: prototype/cogit/rerere.py
"""Conflict resolution memory (COG-020, US-019, closes OQ-008).

Local preference store in .cogit/rerere.json — analogous to git's
rr-cache: mutable, per-repository, never part of immutable history.
Suggestions are surfaced, never auto-applied.
"""

import hashlib
import json
import os
import tempfile

from .canonical import canonical_json_bytes

RERERE_FILE = "rerere.json"


def conflict_fingerprint(conflict) -> str:
    """Orientation-invariant identity of a conflict shape."""
    sides = sorted([sorted(conflict["ours"]), sorted(conflict["theirs"])])
    shape = {"claim": conflict["claim"], "sides": sides, "base": sorted(conflict["base"])}
    return "sha256:" + hashlib.sha256(canonical_json_bytes(shape)).hexdigest()


def load_rerere(cogit_dir) -> dict:
    path = os.path.join(cogit_dir, RERERE_FILE)
    if not os.path.isfile(path):
        return {}
    with open(path, "r", encoding="utf-8") as handle:
        return json.load(handle)


def _save(cogit_dir, data):
    tmp_dir = os.path.join(cogit_dir, "tmp")
    os.makedirs(tmp_dir, exist_ok=True)
    fd, tmp_path = tempfile.mkstemp(prefix="rerere-", dir=tmp_dir)
    try:
        with os.fdopen(fd, "w", encoding="utf-8") as handle:
            json.dump(data, handle, indent=2, sort_keys=True)
            handle.write("\n")
            handle.flush()
            os.fsync(handle.fileno())
        os.rename(tmp_path, os.path.join(cogit_dir, RERERE_FILE))
    finally:
        if os.path.exists(tmp_path):
            os.unlink(tmp_path)


def record_resolution(cogit_dir, conflict, keep, recorded_at):
    """Remember how a conflict was resolved. keep is an assertion id or None (drop)."""
    store = load_rerere(cogit_dir)
    store[conflict_fingerprint(conflict)] = {
        "claim": conflict["claim"],
        "keep": keep,
        "recorded_at": recorded_at,
    }
    _save(cogit_dir, store)


def suggestion_for(cogit_dir, conflict):
    """Stored resolution for this conflict shape, or None."""
    return load_rerere(cogit_dir).get(conflict_fingerprint(conflict))


def forget(cogit_dir, key) -> int:
    """Drop entries by fingerprint or by claim id. Returns removed count."""
    store = load_rerere(cogit_dir)
    doomed = [fp for fp, rec in store.items() if fp == key or rec["claim"] == key]
    for fp in doomed:
        del store[fp]
    if doomed:
        _save(cogit_dir, store)
    return len(doomed)

File: crates/cogit-core/src/rerere.rs
//! Conflict resolution memory (COG-020). Fingerprints MUST match the
//! reference implementation byte-for-byte (shared rerere.json).

use std::fs;
use std::io::Write;
use std::path::Path;

use serde_json::{json, Map, Value};
use sha2::{Digest, Sha256};

use crate::canonical::canonical_json_bytes;
use crate::error::Result;

/// Orientation-invariant identity of a conflict shape.
pub fn conflict_fingerprint(conflict: &Value) -> String {
    let side = |key: &str| -> Vec<String> {
        let mut v: Vec<String> = conflict[key]
            .as_array()
            .into_iter()
            .flatten()
            .filter_map(|x| x.as_str().map(str::to_owned))
            .collect();
        v.sort();
        v
    };
    let mut sides = [side("ours"), side("theirs")];
    sides.sort();
    let shape = json!({
        "claim": conflict["claim"],
        "sides": sides,
        "base": side("base"),
    });
    let digest = Sha256::digest(canonical_json_bytes(&shape).expect("fingerprint shape"));
    let mut out = String::with_capacity(71);
    out.push_str("sha256:");
    for byte in digest {
        out.push_str(&format!("{byte:02x}"));
    }
    out
}

pub fn load_rerere(cogit_dir: &Path) -> Map<String, Value> {
    fs::read_to_string(cogit_dir.join("rerere.json"))
        .ok()
        .and_then(|text| serde_json::from_str::<Value>(&text).ok())
        .and_then(|v| v.as_object().cloned())
        .unwrap_or_default()
}

fn save(cogit_dir: &Path, store: &Map<String, Value>) -> Result<()> {
    let tmp_dir = cogit_dir.join("tmp");
    fs::create_dir_all(&tmp_dir)?;
    let tmp_path = tmp_dir.join(format!("rerere-{}", std::process::id()));
    {
        let mut file = fs::File::create(&tmp_path)?;
        file.write_all(serde_json::to_string_pretty(&Value::Object(store.clone())).unwrap().as_bytes())?;
        file.write_all(b"\n")?;
        file.sync_all()?;
    }
    fs::rename(&tmp_path, cogit_dir.join("rerere.json"))?;
    Ok(())
}

pub fn record_resolution(cogit_dir: &Path, conflict: &Value, keep: Option<&str>, recorded_at: &str) -> Result<()> {
    let mut store = load_rerere(cogit_dir);
    store.insert(
        conflict_fingerprint(conflict),
        json!({
            "claim": conflict["claim"],
            "keep": keep,
            "recorded_at": recorded_at,
        }),
    );
    save(cogit_dir, &store)
}

pub fn suggestion_for(cogit_dir: &Path, conflict: &Value) -> Option<Value> {
    load_rerere(cogit_dir).get(&conflict_fingerprint(conflict)).cloned()
}

/// Drop entries by fingerprint or by claim id. Returns removed count.
pub fn forget(cogit_dir: &Path, key: &str) -> Result<usize> {
    let mut store = load_rerere(cogit_dir);
    let doomed: Vec<String> = store
        .iter()
        .filter(|(fp, rec)| fp.as_str() == key || rec["claim"].as_str() == Some(key))
        .map(|(fp, _)| fp.clone())
        .collect();
    for fp in &doomed {
        store.remove(fp);
    }
    if !doomed.is_empty() {
        save(cogit_dir, &store)?;
    }
    Ok(doomed.len())
}
